design/EP-489.yaml

```diff
@@ -0,0 +1,199 @@
+# EP-489: Secret Lifecycle Management for kagent resources
+
+* Issue: [#489](https://github.com/kagent-dev/kagent/issues/489)
+* Additional Issues:
+  - Issue: [#321](https://github.com/kagent-dev/kagent/issues/321)
+  - Issue: [#575](https://github.com/kagent-dev/kagent/issues/575)
+  - Issue: [#577](https://github.com/kagent-dev/kagent/issues/577)
+
+## Background 
+
+The kagent system creates and manages Kubernetes secrets containing API keys for ModelConfig resources and vector database credentials for Memory resources.
+Currently, when these resources are deleted or updated to no longer use a secret, orphaned secrets remain in the cluster, creating security risks and resource waste.
+
+This enhancement proposes a robust secret lifecycle management system that handles automatic cleanup of unused secrets while avoiding the complexity and race conditions inherent in OwnerReference-based approaches.
+The solution works consistently across all resource creation methods (API, kubectl, Helm).
+
+## Motivation
+
+### Current Problems
+
+1. **Orphaned Secrets**: When ModelConfigs or Memory resources are deleted, their associated secrets remain in the cluster indefinitely
+2. **Update Scenarios**: When resources change providers (e.g., OpenAI --> Ollama, Pinecone --> Mem0), old secrets become orphaned
+3. **Shared Secrets**: Multiple resources can reference the same secret, making ownership unclear
+4. **Manual Secret Creation**: Users can create secrets via kubectl/helm and reference them, bypassing our management
+5. **External Secret Modifications**: Users can delete or modify secrets externally (kubectl/helm), breaking kagent resources that depend on them
+6. **API-Only Solutions Create Inconsistency**: Implementing cleanup only in API handlers would create different behavior between resources created via API vs kubectl/Helm, leading to confusing user experience
+
+### Goals
+
+- Automatically clean up unused secrets when no kagent resources reference them
+- Handle secret cleanup during both deletion and update scenarios  
+- Support shared secrets across multiple ModelConfigs and Memory resources
+- Maintain compatibility with manually created secrets
+- Ensure consistent behavior regardless of resource creation method (API, kubectl, Helm)
+- Avoid race conditions and complex ownership management
+- Follow Kubernetes best practices and security principles
+- Provide extensible solution for future resource types that use secrets
+- Detect and handle external secret modifications (deletion/updates by kubectl/helm)
+
+### Non-Goals
+
+- Managing secrets not related to kagent resources (ModelConfig API keys, Memory credentials, etc.)
+- Cross-namespace secret references (secrets must be in same namespace as referencing resource)
```

Comment:
This unfortunately is not true in our API today, but I would like to change it so that this is true.